plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Completions.hs

```diff
@@ -0,0 +1,656 @@
+      then TopLevel
+      else currentLevel (previousLines pos ls)
+
+-- ----------------------------------------------------------------
+-- Helper Functions
+-- ----------------------------------------------------------------
+
+{- | Takes a position, a list of lines (representing a file)
+  and a map of keywords and returns a keyword context if the
+  previously written keyword matches one in the map
+-}
+getKeyWordContext :: CabalCompletionContext -> [T.Text] -> Map KeyWordName a -> Maybe KeyWordContext
+getKeyWordContext ctx ls keywords = do
+  case lastNonEmptyLineM of
+    Nothing -> Just None
+    Just lastLine' -> do
+      let (whiteSpaces, lastLine) = T.span (== ' ') lastLine'
+      let keywordIndentation = T.length whiteSpaces
+      let cursorIndentation = fromIntegral (pos ^. JL.character) - (T.length $ completionPrefix ctx)
+      -- in order to be in a keyword context the cursor needs
+      -- to be indented more than the keyword
+      if cursorIndentation > keywordIndentation
+        then -- if the last thing written was a keyword without a value
+        case List.find (`T.isPrefixOf` lastLine) (Map.keys keywords) of
+          Nothing -> Just None
+          Just kw -> Just $ KeyWord kw
+        else Just None
+ where
+  pos = completionCursorPosition ctx
+  currentLineM = ls Extra.!? (fromIntegral $ pos ^. JL.line)
+  lastNonEmptyLineM :: Maybe T.Text
+  lastNonEmptyLineM = do
+    cur' <- currentLineM
+    let cur = stripPartiallyWritten $ T.take (fromIntegral $ pos ^. JL.character) cur'
+    List.find (not . T.null . T.stripEnd) $
+      cur : previousLines pos ls
+
+{- | Parse the given set of lines (starting before current cursor position
+  up to the start of the file) to find the nearest stanza declaration,
+  if none is found we are in the top level context.
+-}
+currentLevel :: [T.Text] -> NestingContext
+currentLevel [] = TopLevel
+currentLevel (cur : xs)
+  | Just s <- stanza = Stanza s
+  | otherwise = currentLevel xs
+ where
+  stanza = List.find (`T.isPrefixOf` cur) (Map.keys stanzaKeywordMap)
+
+{- | Returns a CabalCompletionItem with the given starting position
+  and text to be inserted,
+  where the displayed text is the same as the inserted text.
+-}
+makeSimpleCabalCompletionItem :: Range -> T.Text -> CabalCompletionItem
+makeSimpleCabalCompletionItem r txt = CabalCompletionItem txt Nothing r
+
+{- | Returns a CabalCompletionItem with the given starting position,
+  text to be inserted and text to be displayed in the completion suggestion
+-}
+makeCabalCompletionItem :: Range -> T.Text -> T.Text -> CabalCompletionItem
+makeCabalCompletionItem r insertTxt displayTxt =
+  CabalCompletionItem insertTxt (Just displayTxt) r
+
+{- | Get all lines before the given cursor position in the given file
+  and reverse their order to traverse backwards starting from the current position
+-}
+previousLines :: Position -> [T.Text] -> [T.Text]
+previousLines pos ls = reverse $ take (fromIntegral currentLine) ls
```

Comment:
This is linear in the number of lines, right? Are we sure that this is all okay, and we're not accidentally quadratic or something? You might want to use the rope structure that the VFS uses for files, which has more efficient implementations of stuff like this.

Reply:
I switched to the rope structure but unfortunately it does not support any reverse operation.
I opened an (issue)[https://github.com/Bodigrim/text-rope/issues/2] to ask whether there is a better way.

It should be linear though, since in worst case we traverse once from the cursor position to the start of the file.